text_generation/causal_lm/cpp/continuous_batching/library/src/continuous_batching_pipeline.cpp

```diff
@@ -48,10 +48,20 @@ class ContinuousBatchingPipeline::Impl {
 
 
     void _free_non_running_requests() {
-        auto new_end = std::remove_if(m_requests.begin(), m_requests.end(), [] (SequenceGroup::Ptr seq_group) -> bool {
-            return seq_group->has_finished() || seq_group->out_of_memory() || seq_group->handle_dropped();
-        });
-        m_requests.erase(new_end, m_requests.end());
+        std::vector<SequenceGroup::Ptr>::iterator requests_iterator = m_requests.begin();
+        while (requests_iterator != m_requests.end()) {
+            const auto& request = *requests_iterator;
+            if(request->has_finished() || request->out_of_memory() || request->handle_dropped()) {
+                // Assuming there are no running sequences at this point
+                const auto& finished_sequences = request->get_finished_sequences();
```

Comment:
why don't we free all running sequence, but use `get_finished_sequences` ?

Reply:
also, finished should be handled here https://github.com/ilya-lavrenov/openvino.genai/blob/ct-beam-search/text_generation/causal_lm/cpp/continuous_batching/library/src/sampler.hpp#L291, shouldn't they?

Reply:
> why don't we free all running sequence, but use get_finished_sequences ?

I assumed that finished sequence groups have only finished sequences, but for other conditions (like handle_dropped) we may want to clean ongoing sequences, so it indeed might be better and safer to free all of them.

In the sampler they are added to dropped, but they are cleaned on the step() level in CB pipeline. I think we need to do it here because for out-of-memory condition we don't even get to sampler - we drop after scheduler says there are no scheduled tokens.